sgo/types/lookup.go

```diff
@@ -294,6 +293,36 @@ func assertableTo(V *Interface, T Type) (method *Func, wrongType bool, mustOptio
 	return
 }
 
+// typeMustOptional reports whether a type T recursively needs to be wrapped in an optional.
+// If T is Array or Slice, the underlying type will be checked. If T is Optional, it will not
+// need to be wrapped in an optional unless the underlying type is a Map. In that case, further
+// checks are needed for the key and value type of the map.
+func typeMustOptional(T Type) bool {
+	t := T.Underlying()
+	switch t.(type) {
+	case *Pointer, *Map, *Signature, *Chan, *Interface:
+		return true
+	case *Slice, *Array:
```

Comment:
Even if you wrap a struct in an optional it won't save you from a nil on a pointer in a struct, which is the reason why you want to cover also structs, right?

```go
type Foo struct {
    Bar *Bar
}

// ...
myFoo := foo.(?Foo)
myFoo.Bar.DoSomething() // This will crash
```